tempesta_fw/sched.c

```diff
@@ -71,6 +70,166 @@ tfw_sched_get_srv_conn(TfwMsg *msg)
 	return NULL;
 }
 
+/**
+ * Try to reuse last used connection or last used server.
+ */
+static inline TfwSrvConn *
+__try_conn(TfwMsg *msg, TfwStickyConn *st_conn)
+{
+	TfwServer *srv;
+
+	if (unlikely(!st_conn->srv_conn))
+		return NULL;
+
+	if (!tfw_srv_conn_restricted(st_conn->srv_conn)
+	    && !tfw_srv_conn_queue_full(st_conn->srv_conn)
+	    && !tfw_srv_conn_hasnip(st_conn->srv_conn)
+	    && tfw_srv_conn_get_if_live(st_conn->srv_conn))
+	{
+		return st_conn->srv_conn;
+	}
+
+	/* Try to sched from the same server. */
+	srv = (TfwServer *)st_conn->srv_conn->peer;
+
+	return srv->sg->sched->sched_srv_conn(msg, srv);
+}
+
+/**
+ * Find an outgoing connection for client with tempesta sticky cookie.
+ * @sess is not null when calling the function.
+ *
+ * Reuse req->sess->st_conn.srv_conn if it is alive. If not,
+ * then get a new connection for req->sess->srv_conn->peer from
+ * an appropriate scheduler. That eliminates the long generic
+ * scheduling work flow.
+ */
+static inline TfwSrvConn *
+__get_sticky_srv_conn(TfwMsg *msg, TfwHttpSess *sess)
+{
+	TfwStickyConn *st_conn = &sess->st_conn;
+	TfwSrvConn *srv_conn;
+
+	read_lock(&st_conn->lock);
```

Comment:
Isn't it better to use reference counting to make the session private data valid even under failovering?